### PR TITLE
batches: add UI to the preview page to handle UI publication state

### PR DESCRIPTION
**Pull request by LawnGnome**
This is part six of #18277. (This is also the big one.)

This PR extends the functionality of the batch spec preview page to allow for changeset specs without explicit `published` fields to have their publication state controlled from the UI. When one or more changeset specs meet this requirement, the preview page now gains an extra set of checkboxes that can be used to select changeset specs that will be published (either for real or as drafts) when the appropriate option is selected from the dropdown button that replaces the Apply button in the alert.

### Default view with nothing selected

The default apply action still works the same way as before:

image

The dropdown with no changeset specs selected:

image

The dropdown _with_ changeset specs selected:

image

## Implementation notes

The preview page was extremely loosely coupled previously: the alert component (which has the Apply button and also invokes the backend mutation to apply the batch spec) knew very little about the state of any other components on the preview page. Now, though, it needs to be able to track the state of any filters that have been applied, along with knowing which changeset specs have been selected.

Initially, I looked at uplifting the mutation into `BatchChangePreviewPage`, along with the state management. This quickly became unwieldy, though: we have many layers of components on the page, which interact in interesting, often indirect ways (for example, the filter options previously went via URL parameters through `FilteredConnection`).

Instead, I've added two new contexts, which are provided at the `BatchChangePreviewPage` level and then consumed where needed in its descendants:

* `MultiSelectContext`, which tracks the state of the checkboxes, including whether all/visible/individual checkboxes have been selected.
* `BatchChangePreviewContext`, which tracks filter and pagination state, along with overall connection stats that are required across multiple components.

These contexts provide getters and setters for their internal state, which are used by the relevant descendant components.

### Checkbox mania

The checkbox UI is patterned after the bulk operation checkbox UI we have on the actual batch change page. That page couples the checkbox state into the `BatchChangeChangesets` component, which is much more tightly coupled; I used that as inspiration for `MultiSelectContext`, which decouples the checkbox state. Behaviourally, there shouldn't be any significant difference between the two pages.

In the longer term, we may want to consider refactoring `BatchChangeChangesets` to reuse `MultiSelectContext`. It also might not be worth the effort. :man_shrugging:

### CODE REVIEW

**Comment by LawnGnome** [comment]
Opening this up for review, but there are persistent puppeteer failures that I have absolutely no idea how to debug: I can see that two preview tests have failed (bad!), but I can't find any other useful information in the log output (which is massive, and has _so many_ other errors).

Your tax-free donations of help would be greatly appreciated.

**Review by mrnugget — COMMENTED** [review]
Not in-depth review yet (hoping that Erik or Kelli can do a better React/TypeScript review than I can), but just played around with the UI.

What confuses me is the "all", "selected", "visible" distinctions. See video:

https://user-images.githubusercontent.com/1185253/125770796-855884d5-2676-49a0-9a91-70010dededa5.mp4

I would've expected that all non-published changesets would be published as drafts. Looking at the requests though I realised that no publicationStates were sent along. Then I dug into the code to find out why and it looks to me like `"all"` was not selected. But as a user I also don't know how to do that. Is there a "all" button somewhere that shows up if I have more changesets than fit on the page?

If yes then I think (a) the "Publish all [...]" actions shouldn't be shown if the user didn't see that button before or (b) "Publish all [...]" should select all on behalf of the user (I'd prefer b)

File: cmd/frontend/graphqlbackend/batches.graphql [hunk 1]
```diff
@@ -1885,6 +1885,10 @@ type ChangesetApplyPreviewConnectionStats {
     The amount of changesets that are disassociated from the batch change in this operation.
     """
     removed: Int!
+    """
+    The number of changesets that have their publication state controlled from the UI.
+    """
+    uiPublished: Int!
```
[the remark this answers is not in the document]
Comment (eseliger) [reply cmd/frontend/graphqlbackend/batches.graphql:1891]: `uiPublishable`?

File: client/web/src/enterprise/batches/preview/CreateUpdateBatchChangeAlert.tsx [hunk 2]
```diff
@@ -1,99 +1,219 @@
...
+                let publicationStates: ChangesetSpecPublicationStateInput[] | null = null
+                if (action !== CreateUpdateBatchChangeAlertAction.Apply) {
+                    const ids =
+                        selected === 'all'
+                            ? await queryPublishableChangesetSpecs({
+                                  batchSpec: specID,
+                                  ...filters,
+                                  ...pagination,
+                              }).toPromise()
+                            : [...selected]
```
Comment (mrnugget) [inline client/web/src/enterprise/batches/preview/CreateUpdateBatchChangeAlert.tsx:67]: Sooo, after investigating what I described in my comment https://github.com/sourcegraph/sourcegraph/pull/22658#pullrequestreview-707145326 a bit, I think this is more what I would expect:

```suggestion
                    const queryAll =
                        selected === 'all' ||
                        action === CreateUpdateBatchChangeAlertAction.PublishAll ||
                        action === CreateUpdateBatchChangeAlertAction.DraftAll

                    const ids = queryAll
                        ? await queryPublishableChangesetSpecs({
                              batchSpec: specID,
                              ...filters,
                              ...pagination,
                          }).toPromise()
                        : [...selected]
```

That would make the `DraftAll` and `PublishAll` actions work as if the user hit the "Select all" button before.

**Review by mrnugget — COMMENTED** [review]
I dove more into the code and so far everything looks good to me. What I'm worried about and why I don't want to give a final "Approve" yet is the UX of this flow on the preview page that leaves me a bit confused:

1. There is no indication as to what the checkboxes will do here:
image
2. Toggling the checkbox has no effect on the "Actions" column right next to it
3. I only know what the checkbox will do once I hit the dropdown.
4. I can't publish some as draft and some as fully-published. 
5. When I select only draft changesets I still get the "publish as draft" option
6. What if the spec contains other actions - some changesets are closed, some re-opened - and I select changesets to be published: do users really understand that the other changes will also be made?

I guess the underlying problem I have with this flow is that in this context (where we preview what would happen if we don't select something) checking checkboxes and _then_ selecting what the checkboxes mean a bit weird.

Instead I'd prefer to _overwrite_ the spec for certain changesets and then apply that. More concretely: ideally I could select "draft" or "published" next to each changeset and then see its "Action" column update so that the preview is actually a preview of what happens. Then there'd only be a single "Apply spec" button that would apply the preview as we saw it. In that flow there could also be a "Mark all as 'published'" or "Mark all as 'draft'" button.

As I see it we can either

1. go ahead with the flow in this PR and then try to improve it (or not if everybody agrees that what I just wrote is nonsense)
2. take a step back, rethink the flow and redo it

I'm usually a big fan of (1) - merging something and improving it - but when looking at the implementation here I'm wondering whether it isn't too tightly coupled to this flow, making it hard to adapt it to any other flow without deleting large parts of it, in which case it doesn't make sense to merge it, I guess.

In any case: this is just my personal, subjective contribution to a conversation I think we should have, not a hard blocker.

---

I think I also ran into a little bug: going from "draft" to "published" doesn't seem to work. See:

https://user-images.githubusercontent.com/1185253/125787345-9178ade4-6f69-43a2-974e-f078b0b99f14.mp4

**Comment by eseliger** [comment]
I haven't looked too deeply at all of the code yet (and I suspect there isn't anything major), but I've used it a bit locally now and I must say the UX is weird. (Not your fault Adam, I think we didn't find this early enough in the design process, sadly).

The whole preview UI suggests to me now that the entire screen is just about publication status. All the actions themselves are 2nd line now and the massive dropdown at the top makes me think that publish will be the _only_ action that happens. Especially weird on update (see 2nd screenshot, it offers me "Publish selected", but it will actually just do what is listed in the actions column: "reopen" the changesets. No change in publication state whatsoever. [And, to complete this: I think it MUST not change the status of these, because once we start supporting publicationstatus true -> false, not selecting some in the list will mean closing them])

| Create | Update |
| --- | --- |
| image | image |


At the same time, it tells me that it'll publish all, but no actions are required therefor, although there will be actions performed. This lowers my trust in this preview UI _a lot_. Also, that quite detailed UI is quite useless for new batch changes then, since it won't ever display any actions. For update, it at least shows the things that are to be done independently of the publication state change.
Screen Shot 2021-07-15 at 2 25 09 PM
I think a more ideal approach would be to be able to toggle the desired publication state on a per row level (maybe including mass editing functionality by saying "set all to state X"), and then having one column that says desired state: "DRAFT" and internally store that as `map[PreviewRow.ID]PublicationState` and on changes to this map, ask `applyPreview(desiredPublicationStates: myMap)` for an updated preview that then generates a proper plan based on that input and renders the correct actions to be taken then.

I'm not sure what the next steps should be here, but given we have a 2nd entry point into UI publication (bulk operations), maybe we can prioritize that PR first, so we can safely ship this feature this release.

**Comment by rrhyne** [comment]
@eseliger @mrnugget I see what you mean... if I interpret Erik's comment correctly, it seems like we want to use the bulk operation functionality on the preview items to set them all to a certain action, then apply whatever state (the default or the altered state) the user requires.

The bulk operation dropdown is where the 'all as draft', 'all published', etc logic would live to allow users to quickly modify all rows. 

If that is correct, I think we should also add a summary below the text in the banner to summarize what will occur after the bulk operations. Something like:

12 changesets will be published as draft
2 changes will be unpublished
1 no action

**Comment by mrnugget** [comment]
> it seems like we want to use the bulk operation functionality on the preview items to set them all to a certain action, then apply whatever state (the default or the altered state) the user requires.
> 
> The bulk operation dropdown is where the 'all as draft', 'all published', etc logic would live to allow users to quickly modify all rows.

I personally wouldn't use the bulk operations dropdown for that, because that wouldn't allow to overwrite the state for a single changeset: you couldn't say "these 3 should be drafts and these 3 should be published".



> If that is correct, I think we should also add a summary below the text in the banner to summarize what will occur after the bulk operations.

We already have the summary at the top that explains what will happen. Can we use that? Maybe highlight/bold the numbers that were manually overwritten?

image

**Comment by rrhyne** [comment]
I think you could, you'd just have to select the three you wanted to alter and select the action from the dropdown. 

It would make operating on single rows harder, but would users really want to exhaustively click 100 changesets? How often would they be working on just a few changesets?

**Comment by mrnugget** [comment]
Ahh, okay! I think I get it now.

I could select single (or all) changesets and then use a dropdown to say "set action to 'draft'" or "set action to 'publish'", right?

**Comment by LawnGnome** [comment]
After discussion in Slack and our sync, we're going to close this for now and rethink the user story here: it doesn't feel like we're addressing the right concerns with this PR, and we should come back to it with fresher eyes and design this as a separate project after releasing minimal publish from UI support via bulk actions only.

**Pull request closed without merging**